src/etc/wasm32-shim.js

```diff
@@ -0,0 +1,119 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This is a small "shim" program which is used when wasm32 unit tests are run
+// in this repository. This program is intended to be run in node.js and will
+// load a wasm module into memory, instantiate it with a set of imports, and
+// then run it.
+//
+// There's a bunch of helper functions defined here in `imports.env`, but note
+// that most of them aren't actually needed to execute most programs. Many of
+// these are just intended for completeness or debugging. Hopefully over time
+// nothing here is needed for completeness.
+
+const fs = require('fs');
+const process = require('process');
+const buffer = fs.readFileSync(process.argv[2]);
+
+Error.stackTraceLimit = 20;
+
+let m = new WebAssembly.Module(buffer);
+
+let memory = null;
+
+function copystr(a, b) {
+  if (memory === null) {
+    return null
+  }
+  let view = new Uint8Array(memory.buffer).slice(a, a + b);
+  return String.fromCharCode.apply(null, view);
+}
+
+let imports = {};
+imports.env = {
+  // These are generated by LLVM itself for various intrinsic calls. Hopefully
+  // one day this is not necessary and something will automatically do this.
+  fmod: function(x, y) { return x % y; },
```

Comment:
Any reason not to implement these on Rust side? (then they would be "automatic" for any consumer)

Reply:
You mean manually emulating the operation as if they were soft floats? I'm pretty sure that this would be much much slower than calling from wasm to js, executing a native float op, and returning the result to the wasm executable. Also not sure how WASM compilation works but it seems like it is possible to inline that function when the JS runtime JIT compiles it...

Reply:
It wouldn't need soft float; it would just be a call to `fmod` in libm.

Reply:
> to inline that function when the JS runtime JIT compiles it...

It's not because JS and WASM use distinct compilers and runtimes, so every call outside is actually somewhat expensive (just like calls between JS and C++ runtimes).

Reply:
I've discussed this a bit with @nagisa , @rkruppe and @sunfishcode on IRC.

* There is no (fast) hardware instruction for floating point modulo, so you'll *have* to come up with a software solution. However, you can use the hardware instructions for other floating point operations to enable this. This gives you a speedup over pure soft-float implementations. I guess due to this point my point that this code is good is moot (as there will be a non-trivial overhead all the time).
* One can't use libm unless libc is available, as libm is part of libc. The target is named wasm32-unknown-unknown however, it has no libc!
* @nagisa 's math.rs library is built for 100% soft floats, so it is not fast.... @japaric 's m library doesn't seem to include modulo.
* As we don't expose libstd, only libcore, we don't need to support the full set of arithmetic operations

So it seems we'd have to implement the functions ourselves.

Reply:
> math.rs library is built for 100% soft floats, so it is not fast

Might be still worth benchmarking both approaches. It might very well turn out to be still faster than calls to JS side just to modulo two numbers, and even if it's on par, it gives benefit of not forcing each consumer to include these "intrinsics" in their JS consumers.

Reply:
Yeah I'm assuming that one day we can just natively depend on the JS implementation here without implementing our own. Right now though it just requires an opt-in from everyone using it :(

Reply:
Which is why I'm suggesting just including soft emulation of these functions on Rust side. Might be not too bad (and operators that are slow to emulate like `fmod` are not used too frequently anyway).